core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -495,13 +495,39 @@ private[spark] class TaskSetManager(
    * Get the level we can launch tasks according to delay scheduling, based on current wait time.
    */
   private def getAllowedLocalityLevel(curTime: Long): TaskLocality.TaskLocality = {
-    while (curTime - lastLaunchTime >= localityWaits(currentLocalityIndex) &&
-        currentLocalityIndex < myLocalityLevels.length - 1)
-    {
-      // Jump to the next locality level, and remove our waiting time for the current one since
-      // we don't want to count it again on the next one
-      lastLaunchTime += localityWaits(currentLocalityIndex)
-      currentLocalityIndex += 1
+    // remove the emptyList from pendingTasks lazily
+    def hasNonEmptyList(pendingTasks: HashMap[String, ArrayBuffer[Int]]): Boolean = {
```

Comment:
The for-loop will end once there is a list which is not empty, or it will remove all the empty list (not accessed in next time). So I think it will not be expensive even in a big cluster. We also could do it for tasks, then it will fix the false-positive issue.

In short term, this will fix the bug for streaming examples. In long term, this can reduce the latence for streaming (such as updateStateByKey or others needs union).